### PR TITLE
Implement AlphaZero

### PR DESCRIPTION
**Pull request by frechele**
This revision includes:

- Make Alpha Zero
    - Create `MCTSNode`, `AZAgent`, `Network` class
    - Create `ReplayBuffer` class

- Reference
    - https://github.com/maxpumperla/deep_learning_and_the_game_of_go
    - [AlphaZero Paper](https://arxiv.org/abs/1712.01815)
    - [ELF OpenGo Paper](https://arxiv.org/abs/1902.04522)

### CODE REVIEW

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+class MCTSNode(object):
+    def __init__(self, state, prob, parent, action):
+        self.state = state
+        self.parent = parent
+        self.action = action
+
+        self.P = prob
+        self.N = 0
+        self.W = 0
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            new_state = self.state.apply_move(move)
+            prob = probs[move.point.row - 1][move.point.col - 1].item()
+            self.children[move] = MCTSNode(new_state, prob, self, move)
+
+    def expanded(self):
+        return len(self.children) != 0
+    
+    def update(self, value):
+        self.N += 1
+        self.W += value
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            ret[child.action.point.row - 1][child.action.point.col - 1] = child.N / self.N
+
+        return ret.reshape(-1)
+
+    @property
+    def Q(self):
+        if self.N == 0:
+            return 0
+
+        return self.W / self.N
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Function and variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    def update(self, value):
+        self.N += 1
+        self.W += value
+        
+    def inject_noise(self, alpha, eps):
+        noise = np.random.dirichlet([alpha] * len(self.children))
+
+        total_value = 0
+        for i, child in enumerate(self.children.values()):
+            child.P = (1 - eps) * child.P + eps * noise[i]
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            total_value += child.P
+        
+        for child in self.children.values():
+            child.P /= total_value
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        self.N += 1
+        self.W += value
+        
+    def inject_noise(self, alpha, eps):
+        noise = np.random.dirichlet([alpha] * len(self.children))
+
+        total_value = 0
+        for i, child in enumerate(self.children.values()):
+            child.P = (1 - eps) * child.P + eps * noise[i]
+            total_value += child.P
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        
+        for child in self.children.values():
+            child.P /= total_value
+
+    def pi(self):
+        ret = np.zeros((self.state.board.num_rows, self.state.board.num_cols))
+
+        for child in self.children.values():
+            ret[child.action.point.row - 1][child.action.point.col - 1] = child.N / self.N
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+import torch.nn.functional as F
+
+from alphazero import preprocess
+from alphazero.network import Network
+
+from connect5 import agent
+from connect5.types import Player
+from connect5.utils import coords_from_point
+
+USE_CUDA = torch.cuda.is_available()
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    def Q(self):
+        if self.N == 0:
+            return 0
+
+        return self.W / self.N
+
+class AZAgent(agent.Agent):
+    def __init__(self, board_size, state_dict, noise=False, alpha=0.03, eps=0.25, rounds_per_move=1600, puct_init=1.25, puct_base=19652):
+        self.network = Network(board_size)
+        if USE_CUDA:
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        sqrt_total_visit = math.sqrt(max(node.N, 1))
+
+        def score(move):
+            child = node.get_child(move)
+
+            Q = child.Q
+            puct = math.log((1 + child.N + self.puct_base) / self.puct_base) + self.puct_init if c_puct is None else c_puct
+            U = puct * child.P * sqrt_total_visit / (1 + child.N)
+
+            return Q + U
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/mcts.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import torch
+from torch import nn, optim
+
+from alphazero import preprocess
+
+FILTERS = 64
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        self.feature_extraction = nn.Sequential(
+            nn.Conv2d(preprocess.TENSOR_DIM, FILTERS, 3, padding=1),
+            nn.BatchNorm2d(FILTERS),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(FILTERS, FILTERS, 3, padding=1),
+            nn.BatchNorm2d(FILTERS),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(FILTERS, FILTERS, 3, padding=1),
+            nn.BatchNorm2d(FILTERS),
+            nn.ReLU(inplace=True),
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        )
+
+        self.policy_head = nn.Sequential(
+            nn.Conv2d(FILTERS, 2, 1, padding=0),
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+            nn.BatchNorm2d(2),
+            nn.ReLU(inplace=True),
+            Flatten(),
+            nn.Linear(2 * (board_size ** 2), board_size ** 2)
+        )
+
+        self.value_head = nn.Sequential(
+            nn.Conv2d(FILTERS, 1, 1, padding=0),
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/network.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import torch
+
+from connect5.types import Player, Point
+
+PAST_MOVES = 4
+TENSOR_DIM = PAST_MOVES * 2 + 1
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+def StateToTensor(state):
+    rows, cols = state.board.num_rows, state.board.num_cols
+
+    data = torch.zeros(TENSOR_DIM, rows, cols)
+    
+    data[TENSOR_DIM - 1, :, :] = 1 if state.next_player == Player.black else 0
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+    current, opponent = state.next_player, state.next_player.other
+
+    for move in range(min(PAST_MOVES, len(state.previous_states))):
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py]: Variable names should be lowercase by PEP 8

File: 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+                point = Point(col=x+1, row=y+1)
+
+                if state.board.get(point) == current:
+                    data[2 * move + 0, y, x] = 1
+                elif state.board.get(point) == opponent:
+                    data[2 * move + 1, y, x] = 1
+
+        state = state.previous_state
+
+    return data.view(1, TENSOR_DIM, rows, cols)
```
Comment (utilForever) [inline 2 - Examples/2 - Day 2/2 - Alpha Zero/alphazero/preprocess.py]: Variable names should be lowercase by PEP 8

**Review by utilForever — CHANGES_REQUESTED** [review]
스타일 가이드(PEP 8)를 참고하니 몇 군데 고칠 곳이 보이네요. 수정 부탁드립니다.
그리고 VS Code를 쓴다면 `isort` 같은 툴을 사용해 `from` 또는 `import` 순서를 정렬해주세요.
마지막으로 폴더 이름은 `AlphaZero`로 변경 부탁드립니다. (띄어쓰기 X)

**Review by utilForever — APPROVED** [review]
LGTM! :tada:

**Review by 1cekrim — APPROVED** [review]
LGTM. :tada:

**Pull request merged**